### PR TITLE
Change DateTime::timestamp_nanos() to return Option

### LINKED ISSUE #586: DateTime.timestamp_nanos() on valid DateTime panics instead of returning a result/option
While it is documented behaviour, a valid DateTime should not panic while getting a timestamp from it.

### Minimal repro

```rust
let naive = NaiveDateTime::from_timestamp_opt(456456456456, 0).unwrap(); //Does not panic
let datetime = DateTime::<Utc>::from_utc(naive, Utc)); //Still no panic
println!("{}", datetime.timestamp_nanos()); //This panics

```

---

### Potential low-effort fix

```rust
pub fn timestamp_nanos(&self) -> Option<i64> {
    let as_ns = self.timestamp().checked_mul(1_000_000_000);
    match as_ns {
        Some(ns) => Some(ns + i64::from(self.timestamp_subsec_nanos())),
        None => None,
    }
}
```

While testing this potential fix, I uncovered the following points where other unrelated operations on DateTimes call `timestamp_nanos()`, which suggests to me that this might of greater concern than something that can be chalked up as a low importance "formatting bug".

`round.rs: 157`  
`round.rs: 186`

An alternate fix would be to add checked variants of `timestamp_nanos()` and `timestamp_millis()`.

---

As for my use-case, I can easily work around the issue, but this is still a big problem for other use-cases.

### PR DESCRIPTION
**Pull request by Milo123459**
### Thanks for contributing to chrono!

- [x] Have you added yourself and the change to the [changelog]? (Don't worry
      about adding the PR number)
- [x] If this pull request fixes a bug, does it add a test that verifies that
      we can't reintroduce it?

[changelog]: ../CHANGELOG.md

Fixes #586

### CODE REVIEW

**Comment by Milo123459** [comment]
It's failing due to an unrelated change. After looking into it, it is something broken with `bincode`.

**Comment by Milo123459** [comment]
This is breaking because of MSRV. Once #578 gets merged, this can be merged safely.

File: src/datetime.rs [hunk 1]
```diff
@@ -1420,7 +1420,7 @@ pub mod serde {
         where
             S: ser::Serializer,
         {
-            serializer.serialize_i64(dt.timestamp_nanos())
+            serializer.serialize_i64(dt.timestamp_nanos().unwrap())
```
Comment (nickelc) [inline src/datetime.rs:1423]: these unwraps move the panic from `DateTime::timestamp_nanos` to the serde modules.

Comment (Milo123459) [reply src/datetime.rs:1423]: Good catch

**Comment by andy128k** [comment]
Why not `i128`?

**Pull request closed without merging**